packages/server-wallet/src/protocols/close-channel.ts

```diff
@@ -55,70 +43,100 @@ export class ChannelCloser {
           objective.data.targetChannelId,
           tx
         );
-        const nextAction = recordFunctionMetrics(protocol(protocolState), this.timingMetrics);
-
-        if (nextAction) {
-          try {
-            switch (nextAction.type) {
-              case 'SignState':
-                await this.signState(nextAction, protocolState, tx, response);
-                break;
-              case 'CompleteObjective':
-                attemptAnotherProtocolStep = false;
-                await this.completeObjective(objective, protocolState, tx, response);
-                break;
-              case 'Withdraw':
-                await this.withdraw(nextAction, protocolState, tx);
-                break;
-              case 'RequestLedgerDefunding':
-                await this.requestLedgerDefunding(protocolState, tx);
-                break;
-              default:
-                unreachable(nextAction);
-            }
-          } catch (error) {
-            this.logger.error({error}, 'Error handling action');
-            await tx.rollback(error);
+        try {
+          let turnNum;
+          if (!ensureAllAllocationItemsAreExternalDestinations(protocolState)) {
+            attemptAnotherProtocolStep = false;
+          } else if ((turnNum = this.turnNumberToSign(protocolState))) {
+            await this.signState(protocolState, turnNum, tx, response);
+          } else if (this.shouldWithdraw(protocolState)) {
+            await this.withdraw(protocolState, tx);
+          } else if (this.shouldRequestLedgerDefunding(protocolState)) {
+            await this.requestLedgerDefunding(protocolState, tx);
+          } else if (this.shouldCompleteObjective(protocolState)) {
+            attemptAnotherProtocolStep = false;
+            await this.completeObjective(objective, protocolState, tx, response);
+          } else {
+            response.queueChannelState(protocolState.app);
             attemptAnotherProtocolStep = false;
           }
-        } else {
-          response.queueChannelState(protocolState.app);
+        } catch (error) {
+          this.logger.error({error}, 'Error handling action');
```

Comment:
It's not really handling an action anymore, right?

Reply:
fa753ddba